File: client/src/modules/base/buff/NotAccumulateBuff.js
let NotAccumulateBuff = cc.Class.extend({

    ctor: function (){

        this.mapBuff = {};
        this.activeBuff = null;
    },
    getBuffActive: function (clock){

        if(this.activeBuff && !this.activeBuff.checkStatus(clock)){
            this.activeBuff = this.getMaxBuff(clock)
        }

        return this.activeBuff;
    },
    getBuff: function (val){

        return this.mapBuff[val];
    },
    getMaxBuff: function (clock){

        let maxBuff = null;
        let value = 0;
        let keys = Object.keys(this.mapBuff);
        let self = this;
        keys.forEach(function (key){
            if(!self.mapBuff[key].checkStatus(clock)){
                delete self.mapBuff[key]
            }else {
                if(self.mapBuff[key].getValue() > value){
                    value = self.mapBuff[key].getValue();
                    maxBuff = self.mapBuff[key];
                }
            }
        })

        return maxBuff;
    },
    addBuff: function (buff){
        this.mapBuff[buff.getValue()] = buff;
        if(this.activeBuff == null || this.activeBuff.getValue() < buff.getValue()){
            this.activeBuff = buff;
        }
    },
})

File: server/src/battle/model/buff/NotAccumulateBuff.java
package battle.model.buff;

import java.util.ArrayList;
import java.util.HashMap;

/**
 *
 * tao buff khong cong don cac buff cac buff phan biet voi nhau qua key la value
 */
public class NotAccumulateBuff {

    HashMap<Double, Buff> mapBuff = new HashMap<>();
    Buff activeBuff = null;

    /**
     *
     * @param clock thoi gian tren object co buff
     * @return tra ve buff dang duoc kich hoat tren object
     */
    public Buff getBuffActive(int clock){

        if(activeBuff!= null && !activeBuff.checkStatus(clock)){
            activeBuff = getMaxBuff(clock);
        }
        return activeBuff;
    }

    /**
     *
     * @param val gia tri buff
     * @return tra ve buff neu co hoac null
     */
    public Buff getBuff(double val){

        return mapBuff.get(val);
    }

    /**
     *
     * @param buff
     */
    public void addBuff(Buff buff){

        mapBuff.put(buff.getValue(), buff);
        if(activeBuff==null || activeBuff.getValue() < buff.getValue()){
            activeBuff = buff;
        }
    }

    /**
     *
     * @param clock thoi gian tren object co buff
     * @return buff co value lon nhat tren object dang active
     */
    private Buff getMaxBuff(int clock){

        Buff buff = null, maxBuff = null;
        double value = 0;
        ArrayList<Double> removeList = new ArrayList<Double>();
        // duyet qua tat ca cac buff tim buff co value lon nhat dang active
        // va xoa cac buff expired
        for(HashMap.Entry<Double,  Buff> element : mapBuff.entrySet()){

            buff = element.getValue();
            if(!buff.checkStatus(clock)){
                removeList.add(buff.getValue());
            }else {
                if(buff.getValue() > value){
                    value = buff.getValue();
                    maxBuff = buff;
                }
            }
        }
        removeList.forEach((Double buffValue)->{
            mapBuff.remove(buffValue);
        });

        return maxBuff;
    }
}
